### PR TITLE
Feat/add-per-box-embed

### PR DESCRIPTION
**Pull request by ShuaiLYU**
This pull request adds support for returning per-detection (per-box) embedding vectors from the YOLO detection head and propagating them through the inference pipeline to the results. This enables users to optionally attach and access feature embeddings for each detected object, which can be useful for downstream tasks like re-identification or, classification, anomaly scoring. The feature is controlled by a new `embed_boxes` flag.

Key changes include:

**Feature embedding support throughout the pipeline:**

* Added a new `embed_boxes` configuration option to the default configuration and argument parsing, allowing users to enable or disable per-box embeddings. [[1]](diffhunk://#diff-732d61cb4969eff986cec58058000b2949ea3b429a97ca5a1045897e754145c5R73) [[2]](diffhunk://#diff-c02fb3f9e86e97bf85f07211aed56aa04774504a61eeff07a08471bbdea14556R296)
* Modified the YOLO detection head (`Detect`) to optionally output per-box embeddings by splitting the classification branch at the pre-logit layer when `embed_boxes` is enabled. These embeddings are then concatenated to the detection outputs and propagated through postprocessing and NMS. [[1]](diffhunk://#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639R89) [[2]](diffhunk://#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639R156-R160) [[3]](diffhunk://#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639L185-R194) [[4]](diffhunk://#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639L231-R246)

**Inference and results handling:**

* Updated the predictor and inference pipeline to correctly handle the new embeddings, including syncing the `embed_boxes` flag, passing the correct number of classes to NMS, and extracting embeddings from the prediction tensors. [[1]](diffhunk://#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534R306-R311) [[2]](diffhunk://#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23R54-R63) [[3]](diffhunk://#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23L122-R125)
* Extended the `Results` class and its `update` method to store and update per-detection embeddings, ensuring they are accessible alongside other detection outputs. [[1]](diffhunk://#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R254) [[2]](diffhunk://#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R270) [[3]](diffhunk://#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R288-R293) [[4]](diffhunk://#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R338) [[5]](diffhunk://#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R355) [[6]](diffhunk://#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R376-R377)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds optional per-detection embeddings from the YOLO detection head to prediction `Results` objects. 🧠

### 📊 Key Changes
- Adds the `embed_boxes` configuration option to the CLI/config defaults.
- Extracts pre-logit classification-branch features from the detection head when enabled.
- Carries embeddings through inference, NMS, and detection postprocessing while preserving box alignment.
- Exposes embeddings through `Results.embeddings` and includes them in tensor/device operations.
- Synchronizes the embedding flag when reusing predictors.

### 🎯 Purpose & Impact
- Enables downstream tracking, re-identification, similarity search, and other per-object feature workflows. 🎯
- Keeps the feature opt-in, preserving existing prediction behavior by default.
- Adds extra feature memory and computation when enabled, and changes detection outputs to carry embedding columns internally.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShuaiLYU, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message helps with initial review; an Ultralytics engineer will assist with your contribution. Please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee
For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/modules/head.py [hunk 1]
```diff
@@ -152,6 +153,11 @@ def forward_head(
             return {}
         bs = x[0].shape[0]  # batch size
         boxes = torch.cat([box_head[i](x[i]).view(bs, 4 * self.reg_max, -1) for i in range(self.nl)], dim=-1)
+        if self.embed_boxes:  # split cls branch at its last conv to expose the pre-logit embedding
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/head.py:156]: ⚠️ **HIGH**: `embed_boxes` is enabled on every head that inherits `Detect`, but the extra channels are only understood by `DetectionPredictor.construct_result`. This breaks derived tasks: `Segment._inference()` appends mask coefficients after the embeddings, so `SegmentPredictor.construct_result()` passes embeddings plus coefficients to `process_mask`, causing a coefficient/prototype dimension mismatch; pose similarly tries to reshape embeddings plus keypoints as keypoints. Either restrict this path to supported detect heads or update each derived head and predictor to preserve and split embeddings separately.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The standard detection pipeline correctly carries embedding columns through NMS and exposes them on `Results`, but enabling the new global flag regresses inherited segment and pose detection heads because their task-specific extra outputs are no longer parsed at the expected offsets.

💬 Posted 1 inline comment

**Comment by ShuaiLYU** [comment]
Here, I use TSNE to visualize the  bbox embeddings  extracted from the YOLO Head.
```python
# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license
"""Predict over image(s) and plot detections beside a t-SNE of per-box cls-branch embeddings."""

import sys
from math import ceil
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
from sklearn.manifold import TSNE

from ultralytics import YOLO

_C = Path.home() / "../datasets/coco128/images/train2017"  # default: coco128 if present
args = sys.argv[1:] or [str(_C) if _C.exists() else "ultralytics/assets"]
exts = {".jpg", ".jpeg", ".png", ".bmp", ".webp"}
srcs = [(Path(a).rglob("*") if Path(a).is_dir() else [Path(a)] if Path(a).is_file() else Path().glob(a)) for a in args]
imgs = sorted({str(f) for s in srcs for f in s if f.suffix.lower() in exts})

res = [r for r in YOLO("yolo26l.pt").predict(imgs, embed_boxes=True, conf=0.25, verbose=False) if r.embeddings is not None and len(r.boxes)]
emb = np.concatenate([r.embeddings.data.cpu().numpy() for r in res])
cls = np.concatenate([r.boxes.cls.int().cpu().numpy() for r in res])
names, (n, d) = res[0].names, emb.shape
print(f"{len(res)} images, {n} boxes, dim={d}")

# Standardize then t-SNE (raw pre-logit magnitudes overflow sklearn's float32 PCA init)
z = ((emb - emb.mean(0)) / (emb.std(0) + 1e-8)).astype(np.float64)
xy = TSNE(2, perplexity=max(1, min(30, n - 1)), init="random", random_state=0).fit_transform(z) if n > 1 else np.zeros((n, 2))

# Distinct colors for the most frequent classes; the rest drawn as gray "other"
uniq, cnt = np.unique(cls, return_counts=True)
top = uniq[np.argsort(-cnt)][:10]
color = {int(c): plt.get_cmap("tab10")(i % 10) for i, c in enumerate(top)}

fig = plt.figure(figsize=(18, 9))
left, right = fig.subfigures(1, 2, width_ratios=[1, 1.1])
sample = res[:9]
axes = np.atleast_1d(left.subplots(ceil(len(sample) / min(3, len(sample))), min(3, len(sample)))).ravel()
for ax in axes:
    ax.axis("off")
for ax, r in zip(axes, sample):
    ax.imshow(r.plot()[..., ::-1])
left.suptitle(f"Detections (showing {len(sample)} of {len(res)})")

ax = right.subplots()
if (other := ~np.isin(cls, top)).any():
    ax.scatter(xy[other, 0], xy[other, 1], color="lightgray", s=60, label="other", zorder=2)
for c in top:
    m = cls == c
    ax.scatter(xy[m, 0], xy[m, 1], color=color[int(c)], s=90, edgecolors="black", linewidths=0.5, label=f"{names[int(c)]} ({m.sum()})", zorder=3)
ax.set_title(f"t-SNE of box embeddings (images={len(res)}, boxes={n}, dim={d})")
ax.legend(fontsize=8)
ax.grid(True, alpha=0.3)

out = str(Path(__file__).parent / "box_embeddings_tsne.png")
fig.savefig(out, dpi=150, bbox_inches="tight")
print("Saved", out)

```

image
